TombEngine/Objects/TR5/Emitter/Waterfall.cpp

```diff
@@ -0,0 +1,220 @@
+
+            //If targetpos is in a wall, calculate targetpos from the main item coordinates so the particles still spawn.
+            if (pointColl.GetSector().IsWall(part.targetPos.x, part.targetPos.z))
+            {
+                pointColl = GetPointCollision(origin2, item.RoomNumber, item.Pose.Orientation.y, BLOCK(0.0f));
+                relFloorHeight = pointColl.GetFloorHeight() - part.y;
+                origin2 = Geometry::TranslatePoint(Vector3(pos.x, pos.y, pos.z), orient2, BLOCK(0.0f));
+
+                //Set the right height.
+                part.targetPos = Vector4(origin2.x, origin2.y + relFloorHeight, origin2.z, itemNumber);
+
+                //Stop spawning particles if there is a wall directly at the spawnpoint
+                if (pointColl.GetSector().IsWall(pos.x, pos.z))
+                {
+                    part.on = false;
+                    continue;
+                }
+            }
+
+            if (TestEnvironment(ENV_FLAG_WATER, Vector3i(part.targetPos.x, part.targetPos.y, part.targetPos.z), part.roomNumber) ||
+                TestEnvironment(ENV_FLAG_SWAMP, Vector3i(part.targetPos.x, part.targetPos.y, part.targetPos.z), part.roomNumber))
+            {
+                part.targetPos.y = pointColl.GetWaterSurfaceHeight();
+            }
+
+            char colorOffset = Random::GenerateInt(-8, 8);
+
+            part.sR = std::clamp((int)startColor.x + colorOffset, 0, UCHAR_MAX);
+            part.sG = std::clamp((int)startColor.y + colorOffset, 0, UCHAR_MAX);
+            part.sB = std::clamp((int)startColor.z + colorOffset, 0, UCHAR_MAX);
+            part.dR = std::clamp((int)endColor.x + colorOffset, 0, UCHAR_MAX);
+            part.dG = std::clamp((int)endColor.y + colorOffset, 0, UCHAR_MAX);
+            part.dB = std::clamp((int)endColor.z + colorOffset, 0, UCHAR_MAX);
+            part.roomNumber = pointColl.GetRoomNumber();
+            part.colFadeSpeed = 2;
+            part.blendMode = BlendMode::Additive;
+            part.gravity = 120;
+            part.life = part.sLife = WATERFALL_LIFE_MAX;
+            part.fadeToBlack = 0;
+            part.rotAng = Random::GenerateAngle();
+            part.rotAdd = Random::GenerateAngle(ANGLE(-0.1f), ANGLE(0.1f));
+            part.scalar = scale;
+            part.maxYvel = 0;
+            part.sSize = part.size = (item.TriggerFlags < 10 ? Random::GenerateFloat(40.0f, 51.0f) : Random::GenerateFloat(49.0f, 87.0f)) / 2;
+            part.dSize = item.TriggerFlags < 10 ? Random::GenerateFloat(40.0f, 51.0f) : Random::GenerateFloat(98.0f, 174.0f);
+            part.flags = SP_SCALE | SP_DEF | SP_ROTATE;
+        }
+    }
+
+	void SpawnWaterfallMist(const Vector4& pos, int roomNumber, float scalar, float size, const Color& color)
+	{
+		auto& part = *GetFreeParticle();
+
+		auto colorOffset = Color(40.0f, 40.0f, 40.0f); // make constant. Color is Vector 4. 4th value should not be changed.
```

Comment:
`Color` class is used in the wrong way here. `Color` class expects normalized values to be used (from 0.0f to 1.0f). Please convert `color` argument to either `Vector3i` or individual `byte` values, and also don't use `Color` class in similar contexts down the line.